src/android/SecureStorage.java

```diff
@@ -19,24 +35,84 @@
 import org.json.JSONObject;
 import org.json.JSONArray;
 
+import javax.crypto.IllegalBlockSizeException;
+
 public class SecureStorage extends CordovaPlugin {
+    private final ExecutorService threadPool = Executors.newCachedThreadPool();
     private static final String TAG = "SecureStorage";
 
     private static final boolean SUPPORTS_NATIVE_AES = Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP;
     private static final boolean SUPPORTED = Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT;
 
     private static final String MSG_NOT_SUPPORTED = "API 19 (Android 4.4 KitKat) is required. This device is running API " + Build.VERSION.SDK_INT;
     private static final String MSG_DEVICE_NOT_SECURE = "Device is not secure";
+    public static final String MIGRATED_FOR_SECURITY = "_SS_MIGRATED_FOR_SECURITY";
 
     private final Hashtable<String, SharedPreferencesHandler> SERVICE_STORAGE = new Hashtable<String, SharedPreferencesHandler>();
 
     private IntentRequestQueue intentRequestQueue;
 
     @Override
     protected void pluginInitialize() {
+
         super.pluginInitialize();
 
         intentRequestQueue = new IntentRequestQueue(this);
+
+    }
+
+
+    private void securityMigration(CallbackContext callbackContext) throws JSONException {
+        Log.e(TAG, "Migration Started");
+        //transfer all existing items to new table
+        Hashtable<Integer, TransitionValue> transitionTable = new Hashtable<Integer, TransitionValue>();
+        Hashtable<String,Boolean> RSAMap= new Hashtable<String, Boolean>();
+        Enumeration<String> services = SERVICE_STORAGE.keys();
+        while(services.hasMoreElements()){
+            String service = services.nextElement();
+            //initializing rsakeymapper
+            RSAMap.put(service, false);
+            SharedPreferencesHandler handler = SERVICE_STORAGE.get(service);
+            Set<String> keys = handler.keys();
+
+            for(String key : keys){
+                String value = handler.fetch(key);
+                ExecutorResult result = decryptHelper(value, service,callbackContext);
+
+                if(result.type != ExecutorResultType.ERROR){
```

Comment:
This can be destructive. If we're unable to decrypt a key, the customer will lose the value forever. It might happen due to a bug on our side, making this our responsibility.

Maybe we should just cancel the process if this happens, but it may block the customer from using the app if this keeps happening, which is likely. But it gives us an opportunity to fix our bug without losing the data (if that's the case).

What do you think @vmfo22 ?

Reply:
Nice catch @usernuno. Agree with you. We already had an issue in the past because of that. We should try to prevent this kind of situations not to be possible to recover.

Reply:
Actually since there was an error, the value isn't put on the transition table and hence it wont be subject to a overwrite when the values are taken to be reinserted with the new encryption

Reply:
This can still be problematic. If at least one value is not migrated, the application state becomes incoherent. The application logic can be affected. This can generate unexpected behaviours for the user, making the app work incorrectly.

I think we should instead stop the app from running and retry the migration in the next run. If the issue was caused by not having access to the encryption/decryption, everything should work when it is restored. If it is caused by a bug on our side, we can release a new version that fixes it, without risking incorrect behaviours.